File: examples/c/mpmd_roles.c
// Copyright (c) 2026 Bryce M. Westheimer
// SPDX-License-Identifier: BSD-3-Clause

/**
 * @file mpmd_roles.c
 * @brief MPMD Role Manager with DTL C bindings
 *
 * Demonstrates:
 * - dtl_role_manager_create / dtl_role_manager_destroy
 * - dtl_role_manager_add_role / dtl_role_manager_initialize
 * - dtl_intergroup_send / dtl_intergroup_recv for inter-role communication
 *
 * Run:
 *   mpirun -np 4 ./c_mpmd_roles
 */

#include <stdio.h>
#include <stdlib.h>
#include <dtl/bindings/c/dtl.h>

#define CHECK_STATUS(status, msg) \
    do { \
        if (!dtl_status_ok(status)) { \
            fprintf(stderr, "%s: %s\n", msg, dtl_status_message(status)); \
            exit(1); \
        } \
    } while(0)

int main(int argc, char** argv) {
    (void)argc;
    (void)argv;

    dtl_context_t ctx = NULL;
    dtl_status status;

    status = dtl_context_create_default(&ctx);
    CHECK_STATUS(status, "Failed to create context");

    dtl_rank_t rank = dtl_context_rank(ctx);
    dtl_rank_t size = dtl_context_size(ctx);

    if (size < 2) {
        if (rank == 0) printf("This example requires at least 2 ranks.\n");
        dtl_context_destroy(ctx);
        return 1;
    }

    if (rank == 0) {
        printf("DTL MPMD Role Manager (C)\n");
        printf("==========================\n");
        printf("Running with %d ranks\n\n", size);
    }
    dtl_barrier(ctx);

    /* Create role manager */
    dtl_role_manager_t mgr = NULL;
    status = dtl_role_manager_create(ctx, &mgr);
    CHECK_STATUS(status, "Failed to create role manager");

    /* Define roles: 1 coordinator + (size-1) workers */
    status = dtl_role_manager_add_role(mgr, "coordinator", 1);
    CHECK_STATUS(status, "Failed to add coordinator role");

    status = dtl_role_manager_add_role(mgr, "worker", size - 1);
    CHECK_STATUS(status, "Failed to add worker role");

    /* Initialize roles (collective) */
    status = dtl_role_manager_initialize(mgr);
    CHECK_STATUS(status, "Failed to initialize roles");

    /* Query role membership */
    int is_coordinator = 0;
    dtl_role_manager_has_role(mgr, "coordinator", &is_coordinator);

    int is_worker = 0;
    dtl_role_manager_has_role(mgr, "worker", &is_worker);

    printf("Rank %d: coordinator=%d, worker=%d\n",
           rank, is_coordinator, is_worker);
    dtl_barrier(ctx);

    /* Inter-group communication: coordinator sends tasks to workers */
    if (is_coordinator) {
        dtl_size_t worker_size = 0;
        dtl_role_manager_role_size(mgr, "worker", &worker_size);

        printf("\nCoordinator: distributing tasks to %lu workers\n",
               (unsigned long)worker_size);

        for (dtl_rank_t w = 0; w < (dtl_rank_t)worker_size; w++) {
            int32_t task = (w + 1) * 10;
            status = dtl_intergroup_send(mgr, "worker", w, &task, 1,
                                          DTL_DTYPE_INT32, 0);
            if (dtl_status_ok(status)) {
                printf("  Sent task %d to worker %d\n", task, w);
            }
        }

        /* Receive results */
        printf("\nResults:\n");
        for (dtl_rank_t w = 0; w < (dtl_rank_t)worker_size; w++) {
            int32_t result = 0;
            status = dtl_intergroup_recv(mgr, "worker", w, &result, 1,
                                          DTL_DTYPE_INT32, 1);
            if (dtl_status_ok(status)) {
                printf("  Worker %d returned: %d\n", w, result);
            }
        }
    }

    if (is_worker) {
        /* Receive task */
        int32_t task = 0;
        status = dtl_intergroup_recv(mgr, "coordinator", 0, &task, 1,
                                      DTL_DTYPE_INT32, 0);
        CHECK_STATUS(status, "Worker recv failed");

        /* Compute: square the task value */
        int32_t result = task * task;

        dtl_rank_t my_worker_rank = 0;
        dtl_role_manager_role_rank(mgr, "worker", &my_worker_rank);

        printf("Worker %d (global rank %d): received %d, computed %d\n",
               my_worker_rank, rank, task, result);

        /* Send result back */
        status = dtl_intergroup_send(mgr, "coordinator", 0, &result, 1,
                                      DTL_DTYPE_INT32, 1);
        CHECK_STATUS(status, "Worker send failed");
    }

    dtl_barrier(ctx);
    dtl_role_manager_destroy(mgr);
    dtl_context_destroy(ctx);

    if (rank == 0) printf("\nDone!\n");
    return 0;
}

File: examples/python/scripts/mpmd_roles.py
#!/usr/bin/env python3
# Copyright (c) 2026 Bryce M. Westheimer
# SPDX-License-Identifier: BSD-3-Clause

"""
mpmd_roles.py - MPMD Role Manager

Demonstrates:
- dtl.RoleManager for role assignment
- dtl.intergroup_send / dtl.intergroup_recv for inter-role communication
- Coordinator/worker pattern

Run:
    mpirun -np 4 python mpmd_roles.py
"""

import dtl
import numpy as np


def main() -> None:
    with dtl.Context() as ctx:
        if ctx.size < 2:
            if ctx.is_root:
                print("This example requires at least 2 ranks.")
            return

        if ctx.is_root:
            print("DTL MPMD Role Manager (Python)")
            print("================================")
            print(f"Running with {ctx.size} ranks")
            print(f"Coordinator: rank 0")
            print(f"Workers: ranks 1..{ctx.size - 1}\n")
        ctx.barrier()

        # Create role manager
        mgr = dtl.RoleManager(ctx)

        # Define roles
        mgr.add_role("coordinator", 1)
        mgr.add_role("worker", ctx.size - 1)

        # Initialize (collective)
        mgr.initialize()

        is_coordinator = mgr.has_role("coordinator")
        is_worker = mgr.has_role("worker")

        print(f"Rank {ctx.rank}: coordinator={is_coordinator}, worker={is_worker}")
        ctx.barrier()

        if is_coordinator:
            num_workers = mgr.role_size("worker")
            print(f"\nCoordinator: sending tasks to {num_workers} workers")

            # Send tasks
            for w in range(num_workers):
                task = np.array([(w + 1) * 10], dtype=np.int32)
                dtl.intergroup_send(mgr, "worker", w, task, tag=0)
                print(f"  Sent task {task[0]} to worker {w}")

            # Receive results
            print("\nResults:")
            for w in range(num_workers):
                result = dtl.intergroup_recv(mgr, "worker", w,
                                              count=1, dtype=np.int32, tag=1)
                print(f"  Worker {w} returned: {result[0]}")

        if is_worker:
            # Receive task
            task = dtl.intergroup_recv(mgr, "coordinator", 0,
                                        count=1, dtype=np.int32, tag=0)
            worker_rank = mgr.role_rank("worker")

            # Compute: square the value
            result = np.array([task[0] ** 2], dtype=np.int32)
            print(f"Worker {worker_rank} (global rank {ctx.rank}): "
                  f"received {task[0]}, computed {result[0]}")

            # Send result
            dtl.intergroup_send(mgr, "coordinator", 0, result, tag=1)

        ctx.barrier()
        mgr.destroy()

        if ctx.is_root:
            print("\nDone!")


if __name__ == "__main__":
    main()
